### PR TITLE
🪟 🔧 Switch to `pnpm` for package managing

### LINKED ISSUE #21482: Update local webapp development instructions
### Tell us about the documentation you'd like us to add or update.
[Local webapp development.](https://docs.airbyte.com/contributing-to-airbyte/developing-locally#develop-on-airbyte-webapp)
Add `nvm install` command prior to `npm install`

### If applicable, add links to the relevant docs that should be updated
- [Develop on airbyte-webapp](https://docs.airbyte.com/contributing-to-airbyte/developing-locally#develop-on-airbyte-webapp)

### PR DESCRIPTION
**Pull request by timroes**
## What

Switches from using `npm` to `pnpm` for package managing in `airbyte-webapp`.

To use it locally you need to install the exact matching `pnpm` version:

```bash
# <version> must be the exact version from package.json > engines.pnpm
npm install -g pnpm@<version>
# Using it
pnpm i
pnpm start
# ...
```

Closes https://github.com/airbytehq/airbyte/issues/21482

## How

* Replace the `package-lock.json` with the `pnpm-lock.yaml` file.
* Switch Gradle tasks over to use `pnpm` instead of `npm`, also add an additional task that runs before `pnpmInstall` that validates we only have the `pnpm-lock.yml` file and no other lock file (to prevent accidental committing of a new `package-lock.json` file).
* `pnpm` has a patch mechanism built in, so I could remove the nasty `react-virtualize` patch code from Vite and instead use the proper `pnpm` patch mechanism for it.
* Couple of minor required ESLint and TS changes, that were caused by `pnpm` pulling in slightly newer versions.
* Due to differences in some libraries version they also came with new SVGs, so I needed to udpate several snapshot tests.

### CODE REVIEW

File: airbyte-webapp/build.gradle
Comment (timroes) [inline airbyte-webapp/build.gradle:3]: :information_source: Required to have `PnpmTask` available.

File: airbyte-webapp/package.json
Comment (timroes) [inline airbyte-webapp/package.json:12]: :information_source: With `pnpm` the `node_modules/.bin/vite` script is actually a wrapping Bash script not the `js` file of `vite` directly anymore, so we can't call it with `node` anymore, so using `NODE_OPTIONS` to pass down the `-r` parameter to the `node` process that will execute `vite`.

File: airbyte-webapp/src/components/connectorBuilder/useManifestToBuilderForm.ts
Comment (timroes) [inline airbyte-webapp/src/components/connectorBuilder/useManifestToBuilderForm.ts:203]: :information_source: I believe `pnpm` pulled now in a newer version of `orval` which generates the types slightly better, so this isn't needed anymore.

File: airbyte-webapp/src/components/ui/Markdown/remarkAdmonitionsPlugin.ts
Comment (timroes) [inline airbyte-webapp/src/components/ui/Markdown/remarkAdmonitionsPlugin.ts:13]: :information_source: Typings works very fine with the `Root` so no need to pull in `mdast` (which is also the deprecated version of remark) for that.

File: airbyte-webapp/src/components/ui/NextTable/NextTable.tsx
Comment (timroes) [inline airbyte-webapp/src/components/ui/NextTable/NextTable.tsx:30]: :information_source: The type overwrite via `declare` doesn't work with `pnpm` anymore (I don't fully understand why, yet), so moving it into an in place cast instead.

File: airbyte-webapp/.eslintrc.js
Comment (timroes) [inline airbyte-webapp/.eslintrc.js:12]: :information_source: This is with `pnpm` now conflicting with the `react` plugin already being pulled in by `eslint-config-react-app` which we extend above. Since that plugin pulls in the `react` plugin also no need for us to pull it in explicitally.

File: airbyte-webapp/.npmrc
Comment (timroes) [inline airbyte-webapp/.npmrc:2]: :information_source: Required for `pnpm` to execute the `prestart` etc scripts. It doesn't do so by default, but we're using the heavily for Orval compilation.

File: airbyte-webapp/package.json
Comment (timroes) [inline airbyte-webapp/package.json:115]: :information_source: Removed some deprecated package (this was a stub by now, `query-string` include the types itself).

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml
Comment (timroes) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml:129]: :information_source: We'll always need a `type: string` beside the single element `enum`, to make sure Orval can generate the actual type properly, and not render it as `unknown`. cc @flash1293 @lmossman

File: airbyte-webapp/package.json
Comment (timroes) [inline airbyte-webapp/package.json:72]: :information_source: Need to fix that version for now, due to `useBlocker`. This uses unstable code that's no longer there in the most recent react-router-dom version. We should rewrite that code to the newest `unstable_useBlocker` code, but not as part of this PR :D

File: airbyte-webapp/src/components/connection/CatalogTree/next/__snapshots__/BulkEditPanel.test.tsx.snap
Comment (timroes) [inline airbyte-webapp/src/components/connection/CatalogTree/next/__snapshots__/BulkEditPanel.test.tsx.snap:101]: :information_source: Snaptshots updated due to newer versions of those libraries.

**Comment by evantahler** [comment]
One thing we did at $previous-oss-company was add a [`preinstall`](https://github.com/grouparoo/grouparoo/blob/main/package.json#L21) script which [yelled at you](https://github.com/grouparoo/grouparoo/blob/main/bin/preinstall) if you weren't using pnpm.  Folks more-or-less expect that `yarn install` and `npm install` (1) to both work and (2) work equivalently.  We wanted to make it really clear that we _require_ pnpm, and it's not just a suggestion.

**Comment by timroes** [comment]
@evantahler I considered that, the problem is, it's really "slow"... like adding 2-3 seconds. I wanted to avoid that, given that one of the reasons for `pnpm` was the performance benefit (besides more reliable package resolution). That's why I added the gradle build script, so while in theory you can screw up locally (if you don't read the docs), it anyway won't pass CI (in case you didn't notice that you accidentally created a new lock file which didn't exist beforehand). So we'll still require `pnpm` for installation, but unless we see many people stumbling across this, rather prefer not adding that overhead of executing `npx` in `preinstall`.

File: airbyte-webapp/package.json
Comment (lmossman) [inline airbyte-webapp/package.json:12]: We should update the airbyte-webapp/README.md, since it still contains `npm` commands

**Review by lmossman — COMMENTED** [review]
Left some small comments and questions, but overall this seems to work well. Tried clicking around both the Airbyte webapp and the Connector Builder and both seemed to be working as expected.

I'm not going to approve just yet since I had some questions and I'd like some others to take a look at this as well, but let me know if you'd like me to take another look and approve if necessary!

File: airbyte-webapp/package.json
Comment (timroes) [reply airbyte-webapp/package.json:12]: This should be the same as it's currently. This wasn't introduced by `pnpm` but by Vite. They appear multiple times, because Vite forks of workers to run ESLint, TypeScript compiler and Stylelint in, which all run through that same setup code, so you'll see it 4 times. I can take a look (in a different PR), if there's an easy way I can make sure I only log that output if I'm in the "main" process of Vite.

Comment (timroes) [reply airbyte-webapp/package.json:12]: Very good call. I've actually rewritten the README now completely to have less outdated information and just point to our official documentation for development information, so we don't have this duplicated across different places.

File: airbyte-webapp/package.json
Comment (timroes) [reply airbyte-webapp/package.json:72]: I did explicitally NOT mean to update the version. The newer version breaks our `useBlocker` hook, since that was build using internal APIs that are not available in newer versions anymore. So I set a fixed version for now (since `pnpm` would have pulled in a newer one). I've created https://github.com/airbytehq/airbyte/issues/22305 to track rewriting this code and updating.

File: airbyte-webapp/package.json
Comment (timroes) [reply airbyte-webapp/package.json:9]: There is actually no difference in using `npm start` or `pnpm start`. The only difference is for managing libraries, why we make sure we enforce on CI that no one does that wrong and switched accidentally package managers again. Locally I considered (see also Evan's and my comments on this PR) adding a check, but the problem is that it adds "seconds" to the command, which given that `pnpm i` might often only take 2s is a bit frustrating so I decided to get rid of it. Also in addition, it won't really save you much time, since the `preinstall` command actually runs... after install since npm 7 (yeah I know very intuitive), so you'd anyway need to go through the full install first before we print you a warning that you used the wrong command (which you also will see by suddenly a new lock file appearing). And for any of the run commands it's not really needed to check, since there is no difference, so I felt we don't need to add any local checks for this. But please feel free to share opposing opinions.

File: airbyte-webapp/package.json
Comment (josephkmh) [inline airbyte-webapp/package.json:169]: 🥳

**Review by joeykmh — COMMENTED** [review]
Nice! I spun it up and everything I tested seemed to be working properly:
- OSS mode, clicking around & changing source code
- Cloud mode, clicking around & changing source code
- Storybook
- Local build command

Could we also update `airbyte-webapp-e2e-tests/README.md` to reference `pnpm` commands? I don't think we need necessarily to run cypress with `pnpm`, but there are some explicit instructions there about installing and running `airbyte-webapp` with `npm`.

Small note: we'll also want to update the frontend notion page to replace `npm` commands with `pnpm` once this is merged

File: airbyte-webapp/package.json
Comment (lmossman) [reply airbyte-webapp/package.json:72]: Ah okay, I had just misunderstood this then. That makes sense!

File: airbyte-webapp/package.json
Comment (lmossman) [reply airbyte-webapp/package.json:9]: Makes sense to me! Thanks for explaining - I think the approach you've used here sounds good

**Review by lmossman — APPROVED** [review]

File: docs/contributing-to-airbyte/developing-locally.md
Comment (flash1293) [inline docs/contributing-to-airbyte/developing-locally.md:188]: Not a blocker for this PR, but it looks like pnpm ships with a node version manager too, should we just use that one? One tool less

https://pnpm.io/cli/env

Didn't look into it in detail though

**Review by flash1293 — APPROVED** [review]
Tested and everything still seems to work, lgtm.

Should we switch over the e2e test package as well for consistency reasons?

File: docs/contributing-to-airbyte/developing-locally.md
Comment (timroes) [reply docs/contributing-to-airbyte/developing-locally.md:188]: I have looked a bit into it, but currently didn't saw much of a benefit over keep using `nvm` (also we anyway need `npm` first to install `pnpm`, so we either need to guide to `nvm` or a local node installation in our docs).

**Comment by timroes** [comment]
> Should we switch over the e2e test package as well for consistency reasons?

Given we barely update anything in there and don't have much problems with `npm` there, I don't see a reason why... I also don't see a reason why not :shrug: (but would prefer doing it in a separate PR :D)

File: docs/contributing-to-airbyte/developing-locally.md
Comment (flash1293) [reply docs/contributing-to-airbyte/developing-locally.md:188]: You can install pnpm without npm, right? https://pnpm.io/installation

**Comment by flash1293** [comment]
>  I don't see a reason why... I also don't see a reason why not

A user needs to make sure to call either npm or pnpm, depending on whether they are working in e2e or regular webapp (and keep both of them around and up to date)

**Comment by timroes** [comment]
> A user needs to make sure to call either npm or pnpm, depending on whether they are working in e2e or regular webapp (and keep both of them around and up to date)

Problem is, we already have some other modules (the doc system) using yarn anyway, so I feel it's not that bad having `npm` for e2e tests, but I'll change it in a separate PR after this is merged.

**Comment by flash1293** [comment]
> Problem is, we already have some other modules (the doc system) using yarn anyway, so I feel it's not that bad having npm for e2e tests, but I'll change it in a separate PR after this is merged.

Good thing you mention this, that was a little annoying as well :) There's a lot of different technologies in this repo, would be nice to consolidate where possible.

File: airbyte-webapp/package.json
Comment (timroes) [reply airbyte-webapp/package.json:12]: I've opened https://github.com/airbytehq/airbyte/pull/22409 which will improve this, as requested by you.

**Pull request merged**
